apps/gdalalg_vector_check_geometry.cpp

```diff
@@ -0,0 +1,261 @@
+    : GDALVectorPipelineStepAlgorithm(NAME, DESCRIPTION, HELP_URL,
+                                      standaloneStep)
+{
+    AddArg("include-valid", 0,
+           _("Include valid inputs in output, with empty geometry"),
+           &m_includeValid);
+
+    AddArg("geometry-field", 0, _("Name or geometry field to check"),
+           &m_geomField);
+}
+
+#ifdef HAVE_GEOS
+
+class GDALInvalidLocationLayer : public GDALVectorPipelineOutputLayer
+{
+  private:
+    static constexpr const char *ERROR_DESCRIPTION_FIELD = "error";
+
+  public:
+    GDALInvalidLocationLayer(OGRLayer &layer, int srcGeomField, bool skipValid)
+        : GDALVectorPipelineOutputLayer(layer),
+          m_defn(OGRFeatureDefn::CreateFeatureDefn("error_location")),
+          m_geosContext(OGRGeometry::createGEOSContext()),
+          m_srcGeomField(srcGeomField), m_skipValid(skipValid)
+    {
+        m_defn->Reference();
+
+        auto poDescriptionFieldDefn =
+            std::make_unique<OGRFieldDefn>(ERROR_DESCRIPTION_FIELD, OFTString);
+        m_defn->AddFieldDefn(std::move(poDescriptionFieldDefn));
+
+        m_defn->GetGeomFieldDefn(0)->SetSpatialRef(
+            m_srcLayer.GetLayerDefn()
+                ->GetGeomFieldDefn(m_srcGeomField)
+                ->GetSpatialRef());
+    }
+
+    ~GDALInvalidLocationLayer() override;
+
+    int TestCapability(const char *) override
+    {
+        return false;
+    }
+
+    OGRFeatureDefn *GetLayerDefn() override
+    {
+        return m_defn;
+    }
+
+    std::unique_ptr<OGRFeature> CreateFeatureFromLastError()
+    {
+        auto poErrorFeature = std::make_unique<OGRFeature>(m_defn);
+
+        const char *msg = CPLGetLastErrorMsg();
+
+        // Trim GEOS exception name
+        const char *subMsg = strstr(msg, ": ");
+        if (subMsg != nullptr)
+        {
+            msg = subMsg + 2;
+        }
+
+        // Trim newline from end of GEOS exception message
+        if (msg[strlen(msg) - 1] == '\n')
+        {
+            const_cast<char *>(msg)[strlen(msg) - 1] = '\0';
+        }
+
+        poErrorFeature->SetField(ERROR_DESCRIPTION_FIELD, msg);
+
+        CPLErrorReset();
+
+        return poErrorFeature;
+    }
+
+    void TranslateFeature(
+        std::unique_ptr<OGRFeature> poSrcFeature,
+        std::vector<std::unique_ptr<OGRFeature>> &apoOutputFeatures) override
+    {
+        const OGRGeometry *poGeom =
+            poSrcFeature->GetGeomFieldRef(m_srcGeomField);
+        std::unique_ptr<OGRFeature> poErrorFeature;
+
+        if (poGeom)
+        {
+            auto eType = wkbFlatten(poGeom->getGeometryType());
+
+            if (eType != wkbPolygon && eType != wkbMultiPolygon)
+            {
+                CPLErrorOnce(CE_Warning, CPLE_AppDefined,
+                             "Non-polygonal geometry passed to 'gdal vector "
+                             "check-geometry'. Non-polygonal geometries are "
+                             "always valid.");
+            }
```

Comment:
Linestrings can still be not-simple, maybe a new option for the future https://libgeos.org/doxygen/classgeos_1_1operation_1_1valid_1_1IsSimpleOp.html.

Reply:
The GEOS C API doesn't have a way to report the self-intersection location. Maybe I can squeeze that in for the next release.

Reply:
@pramsey